trino-aws-proxy/src/main/java/io/trino/aws/proxy/server/rest/AwsChunkedInputStream.java

```diff
@@ -15,241 +15,213 @@
+                    .entrySet()
+                    .stream()
+                    .filter(entry -> entry.getKey().equalsIgnoreCase("chunk-signature"))
+                    .map(Map.Entry::getValue)
+                    .findFirst();
 
-            if ((separatorIndex + 1) < dataString.length()) {
-                String remainder = dataString.substring(separatorIndex + 1).trim();
-                chunkSignature = Splitter.on(';').trimResults().withKeyValueSeparator('=').split(remainder)
-                        .entrySet()
-                        .stream()
-                        .filter(entry -> entry.getKey().equalsIgnoreCase("chunk-signature"))
-                        .map(Map.Entry::getValue)
-                        .findFirst();
+            if (chunkSignature.isEmpty()) {
+                break;
             }
-            else {
-                chunkSignature = Optional.empty();
+
+            chunkSigningSession.startChunk(chunkSignature.get());
+            bytesRemainingInChunk = chunkSize;
+
+            if (chunkSize == 0) {
+                readEmptyLine();
+                chunkSigningSession.complete();
+                state = State.LAST_CHUNK;
             }
-        }
-        else {
-            chunkSizeString = dataString.trim();
-            chunkSignature = Optional.empty();
-        }
 
-        int chunkSize;
-        try {
-            chunkSize = Integer.parseInt(chunkSizeString, 16);
-        }
-        catch (NumberFormatException e) {
-            throw new IOException("Bad chunk size: " + chunkSizeString);
+            success = true;
         }
+        while (false);
 
-        return new ChunkMetadata(dataString, chunkSize, chunkSignature);
+        if (!success) {
+            throw new IOException("Invalid chunk header: " + header);
+        }
     }
 
-    public void close()
+    private void readEmptyLine()
             throws IOException
     {
-        if (!closed) {
-            try {
-                if (!eof) {
-                    exhaustInputStream(this);
-                }
-            }
-            finally {
-                eof = true;
-                closed = true;
-            }
+        String crLf = readLine();
+        if (!crLf.isEmpty()) {
+            throw new IOException("Expected CR/LF. Instead read: " + crLf);
         }
     }
 
-    @SuppressWarnings("StatementWithEmptyBody")
-    private static void exhaustInputStream(InputStream inStream)
+    private String readLine()
             throws IOException
     {
-        // read and discard the remainder of the message
-        byte[] buffer = new byte[8192];
-        while (inStream.read(buffer) >= 0) {
-            // NOP
+        StringBuilder line = new StringBuilder();
+        while (true) {
+            int i = delegate.read();
+            if (i < 0) {
+                delegateIsDone = true;
+                throw new EOFException("Unexpected end of stream");
```

Comment:
Should we only throw here if `state != State.LAST_CHUNK`?

Reply:
this code should never run if state is `State.LAST_CHUNK`. So, it's best to leave the exception.